Merged gpiotest_mode branch to binder

chrisliebert opened the pull request:
Merged the minimal JSBsim with linking and updated readme.md with windows installation.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hgb2/psas-capstone/14)

<!-- Reviewable:end -->

Commit:
added v1 sensor interface

Commit:
Merged gpiotest_mode branch

cwalexand commented:
Any way to make this work without making users manually change their LD_LIBRARY_PATH environment variable?

test_mode$ cargo run
     Running `target/debug/rcs_sw`
target/debug/rcs_sw: error while loading shared libraries: libJSBSim.so.1: cannot open shared object file: No such file or directory
Process didn't exit successfully: `target/debug/rcs_sw` (exit code: 127)

test_mode$ find /usr/local/lib -name libJSBSim*
/usr/local/lib/libJSBSim.so.1.0.0
/usr/local/lib/libJSBSim.so
/usr/local/lib/libJSBSim.so.1

chrisliebert commented:
The easiest workaround for shared library loading problems would be to execute "export LD_LIBRARY_PATH=/usr/local/lib", or add that line to .bashrc and restart the terminal. Alternatively you could copy or symlink the _.so_ files to the directory where your operating systems currently looks for libraries. That is probably /usr/lib or /usr/lib64. If you choose to install JSBSim to you're system directory instead, you will need to update CMakeLists.txt to find JSBSim (in test_mode/src). It is possible to have CMake search various paths and perform different tasks on different operating environments but it can take a lot of testing to get right.

chrisliebert commented:
Alternatively you could build JSBSim as a static library. You might need to modify the link directive in binder.rs.

chrisliebert commented:
I just verified that you can simply use "cmake -DBUILD_SHARED_LIBS=FALSE" instead when building JSBSim. There is no need to update the link directive in binder.rs to specify kind="static".

BrianBreniser committed:
removed compiler warnings, fixed spelling, formatting

sanayani commented:
another alternative is to put these lines in the build.rs file.  they respectively indicate the library & include path.  s.

println!("cargo:rustc-libdir='/usr/local/lib/'");
println!("cargo:include='/usr/local/include/JSBSim/'");

chrisliebert commented:
The problem @cwalexand was referring to: "unable to load libJSBSim.so" is run-time error so changing the build parameters isn't going to resolve the problem unless libJSBSim is built as a static library.

Commit:
Added note about static linking, and changed test-mode interface to return a new pin each time a value or direction in changed. MyPins has be adjusted to account for these changes.

Commit:
Added loopdata

BrianBreniser committed:
updated agenda

sanayani committed:
update wrapper.h

added externed c functions for most public functions of JSBSim's FGFDMExec class.

sanayani committed:
update wrapper.cpp

add wrappers for most functions in JSBSim's FGFDMExec class.

sanayani committed:
Create wrapper.rs

sanayani committed:
reset binder.rs to work with wrapper.rs

moved the extern block & the unsafe blocks to wrapper.rs.
leaving binder.rs for the higher level init, step (previously loopdata), & close routines.
left placeholder functions in place so that test_mode can compile properly.  the calls to these functions should be updated as noted.

sanayani committed:
update wrapper.rs links

moved the stdc++ & wrapper links from binder.rs.  verify that these are still necessary.

sanayani committed:
removed note re: building jsbsim as static library

do not build jsbsim as a static library.

sanayani committed:
misc edit

sanayani commented:
binder branch:  added major updates to wrapper.h & wrapper.cpp.  separated out the the extern & (the not yet created) wrapper blocks from the previous binder.rs file & put them in new wrapper.rs file.  this should allow us to keep the binder.rs file for just the test_mode functions.  added basic placeholder functions with appropriate signatures.  note that the binder init function should still initialize jsbsim.

sanayani committed:
updated design pseudo code

sanayani committed:
misc edit

sanayani committed:
removed load script

need to verify.

chrisliebert commented on src/test_mode/src/binder.rs:
```diff
@@ -1,52 +1,35 @@
-    }
-
-    // close jsbsim
+    println!("please replace call to binder::terminate with call to binder::close");
```
I think It would be better to remove functions like this entirely and let the compiler notify the developer sine nobody else is using our software at the moment.

chrisliebert commented on src/test_mode/readme.md:
```diff
@@ -20,7 +20,7 @@ sudo make install
 
 ###follow up:
 verify that the installer has placed libJSBSim.so in /usr/local/lib/.
-if you see an error loading libJSBSim.so*, make sure you add /usr/local/lib to LD_LIBRARY_PATH. If you don't want to modify your environment variables, use "-DBUILD_SHARED_LIB=FALSE" instead and JSBSim will be statically linked at build time.
```
Did you mean to remove that note on static linking? I think it is a valid point that @cwalexand brought up.

sanayani commented:
chris:  i left that in there just so that i wouldn't break anything in your code by accident.  if that's not a problem, go ahead & remove both the terminate & loopdata functions.  

with respect to link types, we need to link jsbsim as a shared object & not statically.  please refer to the lines i posted upthread & which tell cargo where to look when it is linking to jsbsim dynamically at run time.  s.

Commit:
Merge branch 'binder' of https://github.com/hgb2/PSAS-Capstone into binder

Commit:
Merged wrapper with binder and added missing } so it will compile

Commit:
Added comment for binder functions
